File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int,int>mpp;
        stack<int>st;
        int n=nums2.size();
        st.push(nums2[n-1]);
        mpp[nums2[n-1]]=-1;
        for(int i=n-2;i>=0;i--)
        {
            if(!st.empty()&&st.top()>nums2[i])
            {
                mpp[nums2[i]]=st.top();
            }
            else
            {
                while(!st.empty()&&st.top()<nums2[i])
                    st.pop();
                 if(st.empty())mpp[nums2[i]]=-1;
                else
                 mpp[nums2[i]]=st.top();
                
            }
             st.push(nums2[i]);
        }
        vector<int>sol;
        for(auto&x:nums1)
        {
            sol.push_back(mpp[x]);
        }
        return sol;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
       HashMap<Integer,Integer>l1=new HashMap<>();
        Stack<Integer>st=new Stack<>();
       
         for(int i=nums2.length-1;i>=0;i--)
         {
                 while(!st.empty()&&st.peek()<nums2[i])st.pop();
                  if(st.empty())l1.put(nums2[i],-1);
                  else l1.put(nums2[i],st.peek());
              st.push(nums2[i]);
             
         }
        int sol[]=new int [nums1.length];
        for(int i=0;i<nums1.length;i++)
        {
            sol[i]=l1.get(nums1[i]);
        }
        return sol;
    }
}